README.md
# KZAsserts - Asserts on roids, test all your assumptions with ease.

[![Version](http://cocoapod-badges.herokuapp.com/v/KZAsserts/badge.png)](http://cocoadocs.org/docsets/KZAsserts)
[![Platform](http://cocoapod-badges.herokuapp.com/p/KZAsserts/badge.png)](http://cocoadocs.org/docsets/KZAsserts)

There are many ways in which we can improve quality of our code-base, Assertions are one of them.
Yet very few people actually use asserts, why is that?

An assert is used to make certain that a specific condition is true, because our apps don't exist in a vacuum  and we often work with other API's (or our own), we need to make sure our assumptions about it are actually valid or we might have really weird bugs.

If our assumption is wrong, assert will crash on the line that checks that condition thus allowing us to find bugs in our code very quickly. 

This is great concept, *crash early and crash often*, it makes it easy to have clean and bug free code.

On the other hand Release build should avoid crashing whenever possible, unless you like 1-star rating?

Because of that normal assertions are **not enough**, you need to make sure your code handles errors even in release builds, if you just strip assertions from release you are still going to crash if your assumption was wrong (eg. after striping asserts you'll have code that mismatches types etc.)!  


How great would it be to assert all assumptions regarding server API responses, so that if backend changes you know it immediately and you can fix your code? So people end up with something like this:  

````
NSParameterAssert([dataFromServer isKindOfClass:[NSDictionary class]]);
if ([dataFromServer isKindOfClass:[NSDictionary class]]) {
  //! create NSError, handle it
}

NSParameterAssert([something isKindOfClass:[NSString class]]);
if ([something isKindOfClass:[NSString class]]) {
  //! create NSError, handle it
}
````
Obviously they could store that condition once and reuse it in assert and if statement, it still sucks!

Now imagine testing your whole response format, that would be so much unnecesary and hard to read code!

That's why I've come up with KZAsserts around 2 years ago, this is how that could can look:

````
AssertTrueOrReturnError([dataFromServer isKindOfClass:[NSDictionary class]]);
AssertTrueOrReturnError([something isKindOfClass:[NSString class]]);
````

This will crash in debug, but in release it will automatically generate NSError for you, it will then return that error object from the current scope. It can also log message to your logger / server / console:
![IMAGE](../blob/master/log.png?raw=true) 

And that's not all, you probably write a lot of async code in your apps? KZAsserts handles that as easily:
````
- (void)downloadFromURL:(NSURL*)url withCompletion:(void (^)(NSData *, NSError *))completionBlock
{
	AssertTrueOrReturnNilBlock([something isKindOfClass:[NSString class]], ^(NSError *error) {
  		completionBlock(nil, error);
	});
	//! ...
}
````

With those kind of macros, you can now assert all your assumptions. 
I'd also encourage you to use Asserts for enforcing API contracts:
eg. if you have a method that downloads NSData * and the only way to get the results is via completion block, you should assert there actually is completion block. Otherwise you are wasting your user battery/network.

## Assert macros

KZAsserts provies following asserts:

````
  AssertTrueOr[X](condition) - if condition fails to be true, on debug builds it will crash by using Assertion, on Release builds it calls error creation and perform specific action. Asserts with block param will execute ^(NSError *){} passed in block with auto-generated NSError.

  AssertTrueOrReturnError
  AssertTrueOrReturnErrorBlock

  AssertTrueOrReturn
  AssertTrueOrReturnBlock

  AssertTrueOrReturnNo
  AssertTrueOrReturnNoBlock

  AssertTrueOrReturnNil
  AssertTrueOrReturnNilBlock

  AssertTrueOrContinue
  AssertTrueOrContinueBlock

  AssertTrueOrBreak
  AssertTrueOrBreakBlock

````

## Installation

KZAsserts is available through [CocoaPods](http://cocoapods.org) for both OSX and iOS, to install
it simply add the following line to your Podfile:

    pod "KZAsserts"

## Supplying your own NSError creation function
If you want to have your own NSError creation function you can just add following line on top of your applicationDidFinishLaunching:
````
[KZAsserts registerErrorFunction:myErrorCreationFunction];
````
You can also change whole format of logging by defining your own KZAMakeError macro.

## Author

Krzysztof Zablocki, [@merowing_](http://twitter.com/merowing_)
[my blog](http://merowing.info)
## License

KZAsserts is available under the MIT license. See the LICENSE file for more info.



CHANGELOG.md
# KZAsserts CHANGELOG

## 0.1.0

Initial release.


Classes/KZAsserts.m
//
//  KZAsserts.h
//  Pixle
//
//  Created by Krzysztof Zabłocki(@merowing_, merowing.info) on 07/2013.
//  Copyright (c) 2013 Pixle. All rights reserved.
//
#import "KZAsserts.h"

const NSUInteger KZAssertFailedAssertionCode = 13143542;

NSError *kza_NSErrorMake(NSString *message, NSUInteger code, NSDictionary *aUserInfo) {
  NSMutableDictionary *userInfo = [aUserInfo mutableCopy];
  userInfo[NSLocalizedDescriptionKey] = message;
  NSError *error = [NSError errorWithDomain:@"info.merowing.internal" code:code userInfo:userInfo];
  NSString *source = error.userInfo[@"Source"] ?: @"";
  NSString *function = error.userInfo[@"Function"] ?: @"";
  
#define RED @"\033[fg214,57,30;"
#define CLEAR "\033[fg;"
#define BLUE @"\033[fg63,126,209;"
#define GREEN @"\033[fg0,244,129;"
  printf("%s", [[NSString stringWithFormat: RED @"KZAsserts" CLEAR BLUE @" %@" CLEAR @" @ " GREEN @"%@" CLEAR RED @" | %@" CLEAR, function, source, message] UTF8String]);

  return error;
}

static TKZAssertErrorFunction function = NULL;

@implementation KZAsserts

+ (void)registerErrorFunction:(TKZAssertErrorFunction)errorFunction
{
  function = errorFunction;
}

+ (TKZAssertErrorFunction)errorFunction
{
  if (!function) {
    [self registerErrorFunction:kza_NSErrorMake];
  }

  return function;
}

@end


Classes/KZAsserts.h
//
//  KZAsserts.h
//  Pixle
//
//  Created by Krzysztof Zabłocki(@merowing_, merowing.info) on 07/2013.
//  Copyright (c) 2013 Pixle. All rights reserved.
//

#ifndef KZAssertsIncluded

#define KZAssertsIncluded true

extern const NSUInteger KZAssertFailedAssertionCode;

typedef NSError *(*TKZAssertErrorFunction)(NSString *message, NSUInteger code, NSDictionary *userInfo);

@interface KZAsserts : NSObject
+ (void)registerErrorFunction:(TKZAssertErrorFunction)errorFunction;

+ (TKZAssertErrorFunction)errorFunction;
@end

#ifndef KZAMakeError
  #define KZAMakeError(condition) KZAsserts.errorFunction([NSString stringWithFormat:@"Condition not satisfied: %s", #condition], KZAssertFailedAssertionCode, @{@"Source" : [NSString stringWithFormat:@"%s:%d", __FILE__, (int)__LINE__], @"Function" : @(__PRETTY_FUNCTION__)}); do{}while(0)
#endif

#define RED @"\033[fg214,57,30;"
#define CLEAR "\033[fg;"
#define BLUE @"\033[fg63,126,209;"
#define GREEN @"\033[fg0,244,129;"

/**
    AssertTrueOr[X](condition) - if condition fails to be true, on debug builds it will crash by using Assertion, on Release builds it calls error creation and perform specific action. Asserts with block param will execute ^(NSError *){} passed in block with auto-generated NSError.
 */
#define AssertTrueOr(condition, action) { BOOL evaluatedCondition = !!(condition); NSCAssert(evaluatedCondition, @"%@", [NSString stringWithFormat:RED @"KZAsserts" CLEAR BLUE @" %s" CLEAR @" @ " GREEN @"%s:%d" CLEAR RED @" | %@" CLEAR, __PRETTY_FUNCTION__, __FILE__, (int)__LINE__, @"Failed: " @#condition]); if (!evaluatedCondition) { NSError *kza_error = KZAMakeError(@#condition); (void)(kza_error); action }} do{} while(0)

#define AssertTrueOrReturnError(condition) AssertTrueOr(condition, return kza_error;)
#define AssertTrueOrReturnErrorBlock(condition, block) AssertTrueOr(condition, block(kza_error); return kza_error;)

#define AssertTrueOrReturn(condition) AssertTrueOr(condition, return;)
#define AssertTrueOrReturnBlock(condition, block) AssertTrueOr(condition, block(kza_error); return;)

#define AssertTrueOrReturnNo(condition) AssertTrueOr(condition, return NO;)
#define AssertTrueOrReturnNoBlock(condition, block) AssertTrueOr(condition, block(kza_error); return NO;)

#define AssertTrueOrReturnNil(condition) AssertTrueOr(condition, return nil;)
#define AssertTrueOrReturnNilBlock(condition, block) AssertTrueOr(condition, block(kza_error); return nil;)

#define AssertTrueOrContinue(condition) AssertTrueOr(condition, continue;)
#define AssertTrueOrContinueBlock(condition, block) AssertTrueOr(condition, block(kza_error); continue;)

#define AssertTrueOrBreak(condition) AssertTrueOr(condition, break;)
#define AssertTrueOrBreakBlock(condition, block) AssertTrueOr(condition, block(kza_error); break;)

#endif